### ISSUE TITLE
Media details' "Copy the wikitext to the clipboard" gives invalid wikitext

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#5075]
I don't think the anchor tag should be there?
```
[[File:SANY R285R excavator in Thailand.jpg|thumb|SANY R285R excavator in Thailand <a href="//commons.wikimedia.org/wiki/File:SANY_R285R_excavator_in_Thailand.jpg#ooui-php-4" title="Edit this at Structured Data on Commons"><img alt="Edit this at Structured Data on Commons" src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/10px-OOjs_UI_icon_edit-ltr-progressive.svg.png" decoding="async" width="10" height="10" style="vertical-align: text-top" srcset="https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/15px-OOjs_UI_icon_edit-ltr-progressive.svg.png 1.5x, https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/20px-OOjs_UI_icon_edit-ltr-progressive.svg.png 2x" data-file-width="20" data-file-height="20"></a>]]
```

**Comment by RitikaPahwa4444** [comment]
Hi @nicolas-raoul, while testing the app, I found that this issues exists only for some images on the app. While I have not been able to figure out the reason for this selective behaviour as of now, I found a fix on stackoverflow:
https://stackoverflow.com/questions/240546/remove-html-tags-from-a-string#:~:text=replace(%22%26%22%2C,line%20breaks%20String%20clean%20%3D%20Jsoup.
I have personally tried the replaceAll() method and it worked well for the image which had tags in its fallback description earlier.

**Comment by nicolas-raoul** [comment]
Seeing the image's page https://commons.wikimedia.org/wiki/File:SANY_R285R_excavator_in_Thailand.jpg I see that the caption does not contain any tag. So I would suggest trying to:
- Identify what API call we are currently getting that string from.
- Identify what API call we could use instead.
That would be more reliable than removing tags afterwards.

**Comment by RitikaPahwa4444** [comment]
Yes, the caption does not contain any tag. But hovering over the edit icon on the image's page shows the same name as found in the tag above:

`title="Edit this at Structured Data on Commons`. 

I think that since most of the other images do not have this icon, they don't have the anchor tag too. 

![Edit Icon](https://user-images.githubusercontent.com/83745993/216568759-559d8feb-4678-4e77-9ec5-63897da74971.jpeg)

Should I check if it is reading the image's description instead of its caption?

**Comment by nicolas-raoul** [comment]
Yes, finding the API call and what part of the API response we use would be a great start :-)

Please post the API call URL when you find it. Thanks!

**Comment by RitikaPahwa4444** [comment]
Thank you for clarifying. Now that I understand the issue better, I will be able to find the cause easily.

**Comment by RitikaPahwa4444** [comment]
I tried looking for the API call and found that the app currently makes use of the following API call URL to fetch the metadata of the image: 

[https://en.wikipedia.org/w/api.php?action=query&format=json&formatversion=2&generator=search&gsrwhat=text&gsrnamespace=6&prop=imageinfo|coordinates&iiprop=url|extmetadata|user&&iiurlwidth=640&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal|Artist|LicenseShortName|LicenseUrl](url)

This is what I found about the prop extmetadata in the official API documentation:

![extmetadata](https://user-images.githubusercontent.com/83745993/217500899-47288722-7e91-400d-b60e-c9f4150e603a.png)

Since the results are HTML formatted, I tried looking for an alternate call that returns only the image description or the image details in an appropriate format, directly or indirectly, but couldn't find anything. However, I found this action which converts HTML into wikitext and is more reliable than stripping the string off to remove the tags:

![flow-parsoid-utils](https://user-images.githubusercontent.com/83745993/217501008-2d142d45-910c-4352-aeb0-2e992e5408ae.png)

**Comment by nicolas-raoul** [comment]
@RitikaPahwa4444 I am getting an error when clicking the URL you posted. One thing, I guess the API call made by the app should be to the Commons server rather than to Wikipedia?

**Comment by RitikaPahwa4444** [comment]
I'm sorry, I should have mentioned that the API call depends on the query at runtime and fetches results accordingly. It needs three parameters: "gsrsearch", "gsrlimit" and "gsroffset". 
Also, while both Commons server  and Wikipedia server have an API call for image info, what I found is that the media that shows up on the explore fragment makes use of the imageInfo and Extmetadata classes in the Wikipedia data client where the base URL corresponds to the one for Wikipedia only. Ideally, it should be the Commons server only and so, I'll try looking for the Commons base URL usages in the project as well.

**Comment by mnalis** [comment]
I'm also experiencing this problem on commons app SOMETIMES in latest f-droid 4.1.0 release (on Android 10). 

- In most cases it works OK (like 00:00-00:13 in this video) where clicking on wikitext button of uploaded image produces correct text like `[[File:Nikola Šubić Zrinski Square.jpg|thumb|Nikola Šubić Zrinski Square (square)]]`

- but in some cases (like 00:16-00:25 in this video) it breaks, and produces invalid text like `[[File:Statue of Nikola Tesla, Zagreb.jpg|thumb|statue of Nikola Tesla, Zagreb <span class="mw-valign-text-top" typeof="mw:File/Frameless"><a href="//commons.wikimedia.org/wiki/File:Statue_of
_Nikola_Tesla,_Zagreb.jpg#ooui-php-4" title="Edit this at Structured Data on Commons"><img alt="Edit this at Structured Data on Commons" src="https://upload.wikimedia.org/wikipedia/commons/th
umb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/10px-OOjs_UI_icon_edit-ltr-progressive.svg.png" decoding="async" width="10" height="10" class="mw-file-element" srcset="https://upload.wikimedia
.org/wikipedia/commons/thumb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/15px-OOjs_UI_icon_edit-ltr-progressive.svg.png 1.5x, https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/OOjs_UI_
icon_edit-ltr-progressive.svg/20px-OOjs_UI_icon_edit-ltr-progressive.svg.png 2x" data-file-width="20" data-file-height="20"></a></span>]]`

Video:

https://github.com/commons-app/apps-android-commons/assets/156656/df640ba4-9f6a-48dd-b871-011f7389e75d

**Comment by mnalis** [comment]
This is the log a day after when I just do the copy/paste with results in incorrect value:  [CommonsAppLogs.zip](https://github.com/commons-app/apps-android-commons/files/12215789/CommonsAppLogs.zip)

And this is a log that seems to contain (among other things) original upload of that problematic `File:Statue of Nikola Tesla, Zagreb.jpg`:  [CommonsAppLogs.zip](https://github.com/commons-app/apps-android-commons/files/12215823/CommonsAppLogs.zip)

Does that help track this down?

**Comment by mnalis** [comment]
FYI, that bug is still present in 5.3.0, e.g.:

https://github.com/user-attachments/assets/d8102d2b-36ee-4fe9-b67b-77044ef05b68

**Comment by rovertrack** [comment]
hey @mnalis  , instead of

 ``` replace("\\<.*?>", "") ```

replacing all the tags and any "<" operators even if they are needed 
why don't we filter / replace the html tags which are present in response  every response are limited 
`<a>`
`<span>`
`<img>`..

**Comment by mnalis** [comment]
> replacing all the tags and any "<" operators even if they are needed

I'm not sure I follow. Ideally should **not** be any HTML tag in that Commons image identifier wikitext.

It should just look like `[[File:Statue of Nikola Tesla, Zagreb.jpg|thumb|statue of Nikola Tesla, Zagreb]]` which includes filename and description (as user entered it in Commons App), and not any HTML tags at all.

But it seems that this `[[Wikitext identifier]]` is constructed not only from `title` field, but also from `ImageDescription.value`, which may be simple text (in majority of cases), but can also be a complex HTML. E.g. `ImageDescription.value` is:

- simple text e.g. [here](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=search&gsrsearch=%22Nikola%20%C5%A0ubi%C4%87%20Zrinski%20Square%20(square)%22&gsrwhat=text&gsrnamespace=6&prop=imageinfo|coordinates&iiprop=url|extmetadata|user): _`Nikola Šubić Zrinski Square (square)`_
- begins normally, but then has extra stuff e.g. [here](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=search&gsrsearch=%22SANY%20R285R%20excavator%20in%20Thailand%22&gsrwhat=text&gsrnamespace=6&prop=imageinfo|coordinates&iiprop=url|extmetadata|user): _`SANY R285R excavator in Thailand <span class="mw-valign-text-top noprint" typeof="mw:File/Frameless"><a href="//commons.wiki…[.png](https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/OOjs_UI_icon_edit-ltr-progressive.svg/20px-OOjs_UI_icon_edit-ltr-progressive.svg.png)" decoding="async" width="10" height="10" class="mw-file-element" data-file-width="20" data-file-height="20"></a></span>`_
- has intermixed useful content with HTML e.g. [here](https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=search&gsrsearch=%22Nikola%20Tesla%20kip%22&gsrwhat=text&gsrnamespace=6&prop=imageinfo|coordinates&iiprop=url|extmetadata|user):  _`Statue of <a href="https://en.wikipedia.org/wiki/Nikola_Tesla" class="extiw" title="w:Nikola Tesla">Nikola Tesla</a> in the …f <a href="https://en.wikipedia.org/wiki/Ivan_Me%C5%A1trovi%C4%87" class="extiw" title="w:Ivan Meštrović">Ivan Meštrović</a>`_

---

That `replace("\\<.*?>", "")` that you mention (BTW is it implemented already? Or just an idea?) would seem to work OK for most of those examples, but not always giving us ideal results. E.g. for examples above it would give:

- _`Nikola Šubić Zrinski Square (square)`_ (GOOD: same untouched text)
- _`SANY R285R excavator in Thailand`_ (GOOD: trims the excess)
-  _`Statue of Nikola Tesla in the …f Ivan Meštrović`_ (PASSABLE: middle of the text is missing, but that is because API didn't return it; and not due to the regexp) -- given that the https://commons.wikimedia.org/wiki/File:Nikola_Tesla_kip.jpg gives full text _`Statue of Nikola Tesla in the street of the same name in Zagreb, work of Ivan Meštrović`_, maybe it should be possible to get **full** correct text using some other API (and then strip HTML tags from it)? 

But **even that case with trimmed text**, it would seem to be **much better** than a load of random HTML crud in a place where it isn't expected (i.e.  image identifier wikitext to be copy/pasted to other apps expecting the image identifier, like e.g. OpenStreetMap apps like [EveryDoor](https://wiki.openstreetmap.org/wiki/Every_Door))...

>  why don't we filter / replace the html tags which are present in response every response are limited
> `<a>`
> `<span>`
> `<img>`..

What would be the benefit on only replacing some tags? I see only extra complexity, and possibility for extra cases we didn't catch to sneak up and would need to fix the same issue again, but no real benefits -- as we **need** to get rid of **all** HTML anyway to get a reasonable plaintext **image identifier** wikitext.
